### ISSUE TITLE
Python API

### ISSUE THREAD

**Issue by domoritz** [uwdata/mosaic#395]
Design and develop a Python API for generating Mosaic specifications. Similar in spirit to [Vega-Altair](https://altair-viz.github.io/), the Python API should enable programmatic construction of a Mosaic JSON specification and enable easy operation with data from Pandas, Polars, DuckDB, etc.

**Comment by domoritz** [comment]
Since https://github.com/uwdata/mosaic/pull/358, Mosaic has a JSON schema for the spec format so we should be able to use https://github.com/vega/altair/tree/main/tools/schemapi to generate the Python API.

**Comment by aeroaks** [comment]
Hi, 
Really amazed by the performance of Mosaic. I can't wait to use it in my daily work to visualise those heavy datasets. 

I looked in the documentation and github and it looks like that currently the best way to use Mosaic from Python is the [Jupyter widget path](https://idl.uw.edu/mosaic/jupyter/). Is that correct?

**Comment by domoritz** [comment]
Yes. I'm planning to work on a Python API for vgplot this semester.

**Comment by aeroaks** [comment]
Great, Count me in if you need testing support. :)

**Comment by domoritz** [comment]
https://pypi.org/project/gosling/ by @manzt looks neat as well and can be an Inspiration. I like how he used Altair as a submodule to get schemapi (which probably should have another release). https://github.com/gosling-lang/gos/tree/main/tools

**Comment by manzt** [comment]
Thanks for the ping. 

If I were starting again, I'd seriously consider using [msgspec](https://jcristharif.com/msgspec/) from @jcrist to build the API layer upon. The separation of encoding/decoding from the base classes is really desirable and makes it so the API layer can be cleaner IMO. Both pydantic and schemapi do not separate the two.

However, in this case I assume you'd like to generate the Python API from some JSON schema. msgspec has the opposite (https://jcristharif.com/msgspec/jsonschema.html), but I wonder with the relatively simpler mosaic JSON schema it would be easier

**Comment by manzt** [comment]
Full disclosure, I've been wanting to move toward msgspec for having widget developers define widgets as  well.

**Comment by domoritz** [comment]
Thanks for the pointer! We should explore that.

> but I wonder with the relatively simpler mosaic JSON schema it would be easier to generate Python API from the JSON schema

What makes mosaic schema simpler? I would like to avoid having to maintain a Python and a TypeScript version of the spec schema. If we can have it in one place and convert to either, that would be best.

**Comment by manzt** [comment]
> What makes mosaic schema simpler?

Sorry, I haven't taken a close look at the spec package... but the thing that makes spec generation "hard" IMO in Altair/Gosling is that TS source makes use of sophisticated TypeScript features like generics and extending base types. 

Having a high-level look at mosiac the types seem to have less inheritance and don't use these features. I believe that would be easier to map to JSON. Could be wrong!

**Comment by domoritz** [comment]
I don't care too much what the source of truth is but I'd like ts and python types. The ts types are being used.

**Comment by manzt** [comment]
> The ts types are being used.

Ah sorry, I wasn't aware. I'd need to explore more to find whats out there.

**Comment by choucavalier** [comment]
a python api would be great. happy to help you on this @domoritz 

maybe we can use pydantic to validate specs? we could use [datamodel-code-generator](https://docs.pydantic.dev/latest/integrations/datamodel_code_generator/) which is able to generate pydantic schemas from JSON schemas

here's a 2-liner self contained way of generating the pydantic models from the latest vgplot json schema:

```bash
wget https://github.com/uwdata/mosaic/raw/refs/heads/main/docs/public/schema/v0.13.0.json
uv run --with datamodel-code-generator datamodel-codegen  --input v0.13.0.json --input-file-type jsonschema --output v0.13.0.py
```

(this takes a few seconds, and generates a 50k LOC Python file)

**Comment by domoritz** [comment]
We have a working prototype at https://github.com/cmudig/mosaic/pull/10. It would be awesome if you could compare your approach to that one. If the current one is better, feel free to send prs for tests, better APIs, docs, etc after coordinating with the students in the pull request. Thanks!

**Comment by dangotbanned** [comment]
Hey @domoritz 👋 

I've been periodically checking in on this issue and thought it would be worth mentioning these PRs:

- [ ] #1007
- [ ] #1057

It looks like quite a lot of progress is being made 🥳

**Comment by domoritz** [comment]
Yes! We are getting very close to merging and making a release.

**Comment by dangotbanned** [comment]
> Yes! We are getting very close to merging and making a release.

Would you be interested in any feedback beforehand?

I'm pleased to see that someone beat me to it on this keyword remapping suggestion 😄 (https://github.com/cmudig/mosaic/commit/33a2d3b934ba93940134bf23b50abecbccc3a381)

**Comment by domoritz** [comment]
Yes! https://github.com/uwdata/mosaic/pull/1007 looks pretty good but I noticed that the generator part is gone and I don't want to manually maintain the API entirely. I started a follow up pull request with API generation in https://github.com/uwdata/mosaic/pull/1067. Yes, a review of the API, the docs, etc would be great. If you have suggestions, feel free to send them as PRs directly against the new branch.

**Comment by domoritz** [comment]
Released in 0.29. Feedback and help are very welcome!
